hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/ZookeeperSubscriptionIdProvider.java

```diff
@@ -0,0 +1,39 @@
+package pl.allegro.tech.hermes.consumers.subscription.id;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.zookeeper.data.Stat;
+import pl.allegro.tech.hermes.api.SubscriptionName;
+import pl.allegro.tech.hermes.common.exception.InternalProcessingException;
+import pl.allegro.tech.hermes.infrastructure.zookeeper.ZookeeperPaths;
+
+import java.util.Optional;
+
+public class ZookeeperSubscriptionIdProvider implements SubscriptionIdProvider {
+
+    private final CuratorFramework curatorFramework;
+    private final ZookeeperPaths zookeeperPaths;
+
+    public ZookeeperSubscriptionIdProvider(CuratorFramework curatorFramework, ZookeeperPaths zookeeperPaths) {
+        this.curatorFramework = curatorFramework;
+        this.zookeeperPaths = zookeeperPaths;
+    }
+
+    @Override
+    public SubscriptionId getSubscriptionId(SubscriptionName name) {
+        return Optional.ofNullable(getZnodeStat(name))
+                .map(Stat::getCzxid)
+                .map(czxid -> SubscriptionId.from(name, czxid))
+                .orElseThrow(() -> new IllegalStateException(
```

Comment:
I'm not sure if it's appropriate use of `IllegalStateException`. 
Documentation says: "Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation."
Maybe could we throw dedicated exception?

Reply:
IMHO it fits actually - when there is no znode related to specific subscription, it means the method was called in inappropriate time, probably after the subscription was removed (or was not created yet, but it's impossible).
I will rethink your suggestion, thanks.